exp/operator/controllers/genericprovider.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// genericProvider a generic type for having ability to write shared function for providers
+type genericProvider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   operatorv1.ProviderSpec   `json:"spec,omitempty"`
+	Status operatorv1.ProviderStatus `json:"status,omitempty"`
+}
+
+func (g *genericProvider) GetConditions() clusterv1.Conditions {
+	return g.Status.Conditions
+}
+
+func (g *genericProvider) SetConditions(conditions clusterv1.Conditions) {
+	g.Status.Conditions = conditions
+}
+
+// DeepCopyObject this is needed only for implementing an interface in order to be able
+// to use conditions library
+func (g *genericProvider) DeepCopyObject() runtime.Object {
+	return nil
+}
```

Comment:
Does the condition setting actually work when you don't have this implemented? 🤔  I wonder if there are unknown consequences of this just returning nil. Is there a way we can actually implement this?

Reply:
I didn't find any issues it can cause and yes condition setting works.

Reply:
I'm +1 to implement this, so we avoid any possible side effect.